include/aws/io/retry_strategy.h

```diff
@@ -0,0 +1,178 @@
+#ifndef AWS_IO_CLIENT_RETRY_STRATEGY_H
+#define AWS_IO_CLIENT_RETRY_STRATEGY_H
+/*
+ * Copyright 2010-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <aws/io/exports.h>
+
+#include <aws/common/atomics.h>
+#include <aws/common/byte_buf.h>
+
+struct aws_retry_strategy;
+struct aws_retry_token;
+struct aws_event_loop_group;
+
+/**
+ * Invoked upon the acquisition, or failure to acquire a retry token. This function will always be invoked if and only
+ * if aws_retry_strategy_acquire_retry_token() returns AWS_OP_SUCCESS. It will never be invoked synchronously from
+ * aws_retry_strategy_acquire_retry_token(). Token will always be NULL if error_code is non-zero, and vice-versa.
+ */
+typedef void(aws_retry_strategy_on_retry_token_acquired_fn)(
+    struct aws_retry_strategy *retry_strategy,
+    int error_code,
+    struct aws_retry_token *token,
```

Comment:
Well we do a sender style here because I don't think we want users dealing with retry strategies once they're working with a token, hence all remaining token functions only take the token pointer.

Reply:
the token does internally point to the retry strategy though.

Reply:
Yeah, we do need the retry_strategy ptr here because `token` will be NULL if you're unable to get the initial token